File: frontend/src/game/upgrades/stat_modifiers/bullet_damage_3.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const BulletDamage3Def: UpgradeDef = {
  id: "bullet_damage_3",
  name: "Sharper Rounds",
  description: "+8 bullet damage",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 8,
  isMultiplier: false,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 99999,
}

export class BulletDamage3 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/bullet_damage_3.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletDamage3(UpgradeImplementation):
    id = 'bullet_damage_3'
    name = 'Sharper Rounds'
    description = '+8 bullet damage'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 8
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
